pretyping/evarconv.ml

```diff
@@ -1110,117 +1211,289 @@ let set_solve_evars f = solve_evars := f
+  let env_evar_unf = evar_env evi in
   let env_evar = evar_filtered_env evi in
   let sign = named_context_val env_evar in
   let ctxt = evar_filtered_context evi in
-  let instance = List.map mkVar (List.map NamedDecl.get_id ctxt) in
-
+  if !debug_ho_unification then
+    (Feedback.msg_debug Pp.(str"env rhs: " ++ print_env env_rhs);
+     Feedback.msg_debug Pp.(str"env evars: " ++ print_env env_evar));
+  let args = Array.map (nf_evar evd) args in
+  let vars = List.map NamedDecl.get_id ctxt in
+  let argsubst = List.map2 (fun id c -> (id, c)) vars (Array.to_list args) in
+  let instance = List.map mkVar vars in
+  let rhs = nf_evar evd rhs in
+  if not (noccur_evar env_rhs evd evk rhs) then raise (TypingFailed evd);
+  (** Ensure that any progress made by Typing.e_solve_evars will not contradict
+      the solution we are trying to build here by adding the problem as a constraint. *)
+  let evd = Evarutil.add_unification_pb (CONV,env_rhs,mkEvar (evk,args),rhs) evd in
+  let evdref = ref evd in
+  let prc env c = print_constr_env env !evdref c in
   let rec make_subst = function
   | decl'::ctxt', c::l, occs::occsl when isVarId evd (NamedDecl.get_id decl') c ->
       begin match occs with
-      | Some _ ->
-        user_err Pp.(str "Cannot force abstraction on identity instance.")
-      | None ->
+      | AtOccurrences loc when not (Locusops.is_all_occurrences loc) ->
+      user_err Pp.(str "Cannot force abstraction on identity instance.")
+      | _ ->
         make_subst (ctxt',l,occsl)
       end
   | decl'::ctxt', c::l, occs::occsl ->
       let id = NamedDecl.get_id decl' in
       let t = NamedDecl.get_type decl' in
       let evs = ref [] in
-      let ty = Retyping.get_type_of env_rhs evd c in
-      let filter' = filter_possible_projections evd c ty ctxt args in
+      let c = nf_evar evd c in
+      (* ty is in env_rhs now *)
+      let ty = replace_vars argsubst t in
+      let filter' = filter_possible_projections !evdref c (nf_evar evd ty) ctxt args in
       (id,t,c,ty,evs,Filter.make filter',occs) :: make_subst (ctxt',l,occsl)
   | _, _, [] -> []
   | _ -> anomaly (Pp.str "Signature or instance are shorter than the occurrences list.") in
-
-  let rec set_holes evdref rhs = function
-  | (id,_,c,cty,evsref,filter,occs)::subst ->
-      let set_var k =
-        match occs with
-        | Some Locus.AllOccurrences -> mkVar id
-        | Some _ -> user_err Pp.(str "Selection of specific occurrences not supported")
-        | None ->
-        let evty = set_holes evdref cty subst in
-        let instance = Filter.filter_list filter instance in
-        let evd = !evdref in
-        let (evd, ev) = new_evar_instance sign evd evty ~filter instance in
-        evdref := evd;
-        evsref := (fst (destEvar !evdref ev),evty)::!evsref;
-        ev in
-      set_holes evdref (apply_on_subterm env_rhs evdref set_var c rhs) subst
+  let fixed = ref Evar.Set.empty in
+  let rec set_holes env_rhs evdref rhs = function
```

Comment:
I think it would make the code simpler to understand if this function would take in input an evd, and use an evdref only locally to call other functions that use the imperative API

Reply:
Indeed. I could avoid it entirely in second_order_matching, there remains cases in apply_on_subterm and thin_evars which make use of map_constr_with_binders that doesn't naturally allow side effects on the evar_map.